sway/input/cursor.c

```diff
@@ -1284,3 +1392,111 @@ const char *get_mouse_button_name(uint32_t button) {
 	}
 	return name;
 }
+
+static void warp_to_constraint_cursor_hint(struct sway_cursor *cursor) {
+	struct wlr_pointer_constraint_v1 *constraint = cursor->active_constraint;
+
+	if (constraint->current.committed &
+			WLR_POINTER_CONSTRAINT_V1_STATE_CURSOR_HINT) {
+		double sx = constraint->current.cursor_hint.x;
+		double sy = constraint->current.cursor_hint.y;
+
+		struct sway_view *view = view_from_wlr_surface(constraint->surface);
+		struct sway_container *con = view->container;
+
+		double lx = sx + con->content_x - view->geometry.x;
+		double ly = sy + con->content_y - view->geometry.y;
+
+		wlr_cursor_warp(cursor->cursor, NULL, lx, ly);
+	}
+}
+
+void handle_constraint_destroy(struct wl_listener *listener, void *data) {
+	struct sway_pointer_constraint *sway_constraint =
+		wl_container_of(listener, sway_constraint, destroy);
+	struct wlr_pointer_constraint_v1 *constraint = data;
+	struct sway_seat *seat = constraint->seat->data;
+	struct sway_cursor *cursor = seat->cursor;
+
+	wl_list_remove(&sway_constraint->destroy.link);
+
+	if (cursor->active_constraint == constraint) {
+		warp_to_constraint_cursor_hint(cursor);
+
+		if (cursor->constraint_commit.link.next != NULL) {
+			wl_list_remove(&cursor->constraint_commit.link);
+		}
+		wl_list_init(&cursor->constraint_commit.link);
+		cursor->active_constraint = NULL;
+	}
+
+	free(sway_constraint);
+}
+
+void handle_pointer_constraint(struct wl_listener *listener, void *data) {
```

Comment:
This can be `static`, and be set in `sway_cursor_constrain` near `wl_signal_add`. Makes code easier to follow.